src/heartbeats.cpp

```diff
@@ -0,0 +1,605 @@
+    {
+        mPendingUps = pendingUps;
+        updateLastActionTime();
+    }
+}
+
+void HeartBeatBackupInfo::setStatus(const Status &status)
+{
+    if (mStatus != status)
+    {
+        mStatus = status;
+        updateLastActionTime();
+    }
+}
+
+void HeartBeatBackupInfo::setLastAction(const m_time_t &lastAction)
+{
+    mLastAction = lastAction;
+}
+
+void HeartBeatBackupInfo::updateLastActionTime()
+{
+    setLastAction(m_time(nullptr));
+}
+
+void HeartBeatBackupInfo::setLastBeat(const m_time_t &lastBeat)
+{
+    mLastBeat = lastBeat;
+}
+
+m_time_t HeartBeatBackupInfo::lastBeat() const
+{
+    return mLastBeat;
+}
+
+void HeartBeatBackupInfo::onCommandToBeDeleted(Command *command)
+{
+    if (mRunningCommand == command)
+    {
+        mRunningCommand = nullptr;
+    }
+}
+
+MegaBackupMonitor::MegaBackupMonitor(MegaClient *client)
+    : mClient(client)
+{
+}
+
+MegaBackupMonitor::~MegaBackupMonitor()
+{
+}
+
+void MegaBackupMonitor::digestPutResult(handle backupId)
+{
+    bool needsAdding = true;
+
+    // get the tag from queue of pending puts
+    assert(mPendingBackupPuts.size());
+    auto syncTag = mPendingBackupPuts.front();
+    mPendingBackupPuts.pop_front();
+
+    if (ISUNDEF(backupId))
+    {
+        LOG_warn << "Received invalid id for sync with tag: " << syncTag;
+        needsAdding = false;
+    }
+
+    // set heartBeat ID
+
+    if (!ISUNDEF(backupId) && mHeartBeatedSyncs.find(syncTag) != mHeartBeatedSyncs.end())
+    {
+        mHeartBeatedSyncs[syncTag]->setBackupId(backupId);
+        needsAdding = false;
+    }
+
+    if (needsAdding)
+    {
+        for (const auto &hBPair : mHeartBeatedSyncs)
+        {
+            if (hBPair.second->backupId() == backupId)
+            {
+                needsAdding = false;
+                break;
+            }
+        }
+
+        if (needsAdding)
+        {
+            //create new HeartBeatSyncInfo
+            mHeartBeatedSyncs[syncTag] = std::make_shared<HeartBeatBackupInfo>(syncTag, backupId);
+        }
+    }
+
+    // store the id in the sync configuration
+    if (!ISUNDEF(backupId))
+    {
+        mClient->updateSyncBackupId(syncTag, backupId);
+    }
+
+    // handle pending updates
+    const auto &pendingSyncPair = mPendingSyncUpdates.find(syncTag);
+    if (pendingSyncPair != mPendingSyncUpdates.end())
+    {
+        if (!ISUNDEF(backupId))
+        {
+            updateSyncInfo(backupId, pendingSyncPair->second.get());
+        }
+        else
+        {
+            LOG_warn << "discarding heartbeat update for pending sync: no valid id received for sync: " << pendingSyncPair->second->getLocalFolder();
+
+        }
+        mPendingSyncUpdates.erase(pendingSyncPair);
+    }
+}
+
+int MegaBackupMonitor::getHBState(MegaSync *sync)
+{
+    if (sync->isTemporaryDisabled())
+    {
+        return MegaBackupMonitor::State::TEMPORARY_DISABLED;
+    }
+    else if (sync->isActive())
+    {
+        //TODO: consider use case: paused, if transfers are paused!?
```

Comment:
Are you suggesting a new state `PAUSED`?? Perhaps it's useful. Should we ask André about APIs support?

Reply:
`Paused` was stated in the specs. Although I haven't added it since there is no `pause` referred to a sync. We could infer it from transfers pauses, but I'm not 100% that's what's wanted ...

Reply:
addresed in later commits